### PR TITLE
Refactor cluster error handling

### PR DESCRIPTION
**Pull request by jaymell**
This make some changes to how errors are handled in the redis cluster clients, most notably:
* Remove exclusions -- in practice, exclusions create a lot of unnecessary connection rebuilding/slot refreshes when a failed command results in a `MOVED` error.
* Attempt to refresh failed connections when IO errors occur instead of adding exclusions and ultimately refreshing slots due to the above-described scenario
* Only retry given a subset of error types, otherwise return error immediately.

### CODE REVIEW

File: redis/tests/support/mock_cluster.rs [hunk 1]
```diff
@@ -121,6 +121,34 @@ pub fn respond_startup(name: &str, cmd: &[u8]) -> Result<(), RedisResult<Value>>
     }
 }
 
+pub fn respond_startup_two_nodes(name: &str, cmd: &[u8]) -> Result<(), RedisResult<Value>> {
```
Comment (nihohit) [inline redis/tests/support/mock_cluster.rs:124]: instead of this, maybe just pull https://github.com/redis-rs/redis-rs/pull/843/files#diff-ae2ede8b12dd85b78d0a380121b8b51caedf3f46452a564b18454dc8b5fe354c ? That way we won't have logic conflicts later.

File: redis/src/cluster_routing.rs [hunk 2]
```diff
@@ -210,7 +210,39 @@ impl<'a> IntoIterator for &'a SlotAddrs {
     }
 }
 
-pub(crate) type SlotMap = BTreeMap<u16, SlotAddrs>;
+#[derive(Debug, Default)]
+pub(crate) struct SlotMap(BTreeMap<u16, SlotAddrs>);
```
Comment (nihohit) [inline redis/src/cluster_routing.rs:214]: IMO this is a self contained change, that can be moved to a separate PR. Merging this early will make the review process for the larger PR faster, since it will reduce the amount of code to review.

File: redis/src/types.rs [hunk 3]
```diff
@@ -581,6 +581,32 @@ impl RedisError {
         };
         Self { repr }
     }
+
+    #[cfg(feature = "cluster")] // Used to avoid "unused method" warning
+    pub(crate) fn is_retryable(&self) -> bool {
```
Comment (nihohit) [inline redis/src/types.rs:586]: IMO this should be pub, not pub(crate) - this is usable to anyone who wants to implement a client over redis-rs. 
This is also a self contained change, that can be moved to a separate PR. Merging this early will make the review process for the larger PR faster, since it will reduce the amount of code to review.

**Review by nihohit — COMMENTED** [review]
this was a high-level review, I didn't go into each line.
I think it's a great change, I'll be happy to see it land soon.

**Review by nihohit — COMMENTED** [review]
this was a high-level review, I didn't go into each line.
I think it's a great change, I'll be happy to see it land soon.

File: redis/src/types.rs (hunk 3 above)
Comment (jaymell) [reply redis/src/types.rs:592]: I think I'd like to leave this in the PR since integrating it into both clients will require a lot of the larger refactoring that's in here. I'll admit this is PR is a bit more unwieldy than I'd hoped and didn't quite end up where I intended it to :sweat_smile:. 

I'm not opposed to making `is_retryable` public, but for the moment I'm not sure whether all of the variants are handled properly so thought I'd avoid exposing it for now.

File: redis/tests/test_cluster.rs [hunk 4]
```diff
@@ -433,3 +439,71 @@ fn test_cluster_replica_read() {
...
+                completed.fetch_add(1, Ordering::SeqCst);
+                Err(parse_redis_value(b"-ERR mock\r\n"))
+            }
+        },
+    );
+
+    let value = cmd("GET").arg("test").query::<Option<i32>>(&mut connection);
+
+    match value {
+        Ok(_) => panic!("This is not ok, man"),
```
Comment (nihohit) [inline redis/tests/test_cluster.rs:502]: same

File: redis/src/cluster_async/mod.rs [hunk 5]
```diff
@@ -282,38 +287,42 @@ where
...
-                        // Refresh slots and request again.
-                        request.info.excludes.clear();
-                        return Next::Err {
-                            request: this.request.take().unwrap(),
-                            error: err,
-                        }
-                        .into();
-                    } else if error_code == "TRYAGAIN" || error_code == "CLUSTERDOWN" {
+                match err.kind() {
+                    // TODO: Proper Ask implementation
```
Comment (nihohit) [inline redis/src/cluster_async/mod.rs:291]: please add what's missing from the current implementation

File: redis/src/cluster_async/mod.rs [hunk 6]
```diff
@@ -496,47 +513,26 @@ where
...
+            let route = info.route.as_ref().unwrap();
+            if let Some(addr) = self.slots.slot_addr_for_route(route) {
+                let addr = addr.to_string();
+                if let Some(conn) = self.connections.get(&addr) {
+                    (addr, conn.clone())
+                } else {
+                    get_random_connection(&self.connections)
+                }
+            } else {
+                get_random_connection(&self.connections)
```
Comment (nihohit) [inline redis/src/cluster_async/mod.rs:534]: minor:
this 3-fold do {} else {get_random_connection}

makes me think that the structure of lines 522-534 can be terser. Maybe something like this:

```rs
 let (addr, conn) = info.route
   .and_then(|route| self.slots.slot_addr_for_route(route)
   .and_then(|addr| {
      let addr = addr.to_string();
      self.connections.get(&addr).map(|conn| (addr, conn))
   }).unwrap_or_else(||get_random_connection(&self.connections));
```

File: redis/src/types.rs [hunk 7]
```diff
@@ -587,6 +587,32 @@ impl RedisError {
...
+        match self.kind() {
+            ErrorKind::ExecAbortError => true,
```
[the remark this answers is not in the document]
Comment (jaymell) [reply redis/src/types.rs:594]: `ExecAbortError` definitely not.

`I'm not clear on `ExtensionError` -- it's a bit of a catch-all.

`Readonly` and `ClusterDown` -- I'm also not sure but thought they might potentially arise due to intermittent cluster flakiness?

[the remark this answers is not in the document]
Comment (nihohit) [reply redis/src/types.rs:594]: > I'm not clear on `ExtensionError` -- it's a bit of a catch-all.

IMO without additional information, we can't know whether this is a transitory error or not. In such a situation, we should pass if to the user, and let them decide, given their knowledge of what modules they use and how the modules behave, whether to retry or not.

> Readonly and ClusterDown -- I'm also not sure but thought they might potentially arise due to intermittent cluster flakiness?

fair enough.

File: redis/src/cluster_async/mod.rs (hunk 6 above)
Comment (jaymell) [reply redis/src/cluster_async/mod.rs:534]: Yep, looks great!

File: redis/src/cluster.rs [hunk 8]
```diff
@@ -505,41 +503,41 @@ where
...
-
-                    excludes.insert(addr);
-
-                    let connections = self.connections.borrow();
-                    if excludes.len() >= connections.len() {
-                        return Err(err);
+                        ErrorKind::IoError => {
+                            if *self.auto_reconnect.borrow() {
+                                if let Ok(mut conn) = self.connect(&addr) {
+                                    // FIXME: Duplicated from `refresh_slots`
```
[the remark this answers is not in the document]
Comment (jaymell) [reply redis/src/cluster.rs]: Moved into `ClusterConnection::connect`. I don't see a reason not to consolidate there.

File: redis/src/cluster_async/mod.rs (hunk 5 above)
Comment (jaymell) [reply redis/src/cluster_async/mod.rs:291]: I'm working on a followup PR that should add the implementation shortly, so I think this is ok for now.

File: redis/src/cluster_async/mod.rs [hunk 9]
```diff
@@ -398,14 +407,32 @@ where
...
+    fn refresh_connections(
+        &mut self,
+        addrs: Vec<String>,
+    ) -> impl Future<Output = ConnectionMap<C>> {
+        let mut connections = mem::take(&mut self.connections);
+        let params = self.cluster_params.clone();
+        async move {
+            for addr in addrs {
+                let conn =
+                    Self::get_or_create_conn(&addr, connections.remove(&addr), params.clone())
```
[the remark this answers is not in the document]
Comment (jaymell) [reply redis/src/cluster_async/mod.rs]: I went ahead and parallelized it.

File: redis/src/types.rs [hunk 10]
```diff
@@ -587,6 +587,32 @@ impl RedisError {
         };
         Self { repr }
     }
+
+    #[cfg(feature = "cluster")] // Used to avoid "unused method" warning
+    pub(crate) fn is_retryable(&self) -> bool {
```
[the remark this answers is not in the document]
Comment (jaymell) [reply redis/src/types.rs:592]: I think it's a good idea but would need to consider further. I think there's probably still larger refactoring needed around retries generally (i.e., in non-clustered use cases as well), so would probably defer this for now.

[the remark this answers is not in the document]
Comment (nihohit) [reply redis/src/types.rs:592]: fair enough. Would you mind adding a TODO, so we won't forget?

File: redis/src/cluster_routing.rs [hunk 11]
```diff
@@ -237,10 +237,6 @@ impl SlotMap {
     pub fn values(&self) -> std::collections::btree_map::Values<u16, SlotAddrs> {
         self.0.values()
     }
-
-    pub fn len(&self) -> usize {
```
[the remark this answers is not in the document]
Comment (jaymell) [reply redis/src/cluster_routing.rs]: Added back in -- was temporarily unused during refactoring :smirk:

**Review by nihohit — APPROVED** [review]

**Comment by nihohit** [comment]
LGTM
